ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java

```diff
@@ -269,7 +273,15 @@ public static Handler requireAuth(Class<? extends Client<?, ?>> clientType, Auth
    */
   public static Promise<UserProfile> login(Context ctx, Class<? extends Client<?, ?>> clientType) {
     return userProfile(ctx)
-      .route(p -> !p.isPresent(), p -> initiateAuthentication(ctx, clientType))
+      .flatMap(p -> {
+        if (!p.isPresent() && DirectClient.class.isAssignableFrom(clientType)) {
+          return performDirectAuthentication(ctx, clientType);
+        } else {
+          return Promise.value(p);
+        }
+      })
+      .route(p -> !p.isPresent() && DirectClient.class.isAssignableFrom(clientType), p -> ctx.clientError(401))
+      .route(p -> !p.isPresent() && IndirectClient.class.isAssignableFrom(clientType), p -> initiateAuthentication(ctx, clientType))
```

Comment:
I'm not sure if there's a neater way to handle this with `Promise`. Essentially what it's doing is in the `flatMap` attempting to get a `UserProfile` directly from the credentials (token or basic auth header). Because I want to populate the promise I couldn't do it all in `route` so I've got an awkward split of `flatMap` and `route` with a similar conditional.